mail_activity_reminder/models/mail_activity.py

```diff
@@ -0,0 +1,132 @@
+        compute_sudo=True,
+        store=True,
+    )
+    last_reminder_local = fields.Datetime(
+        string='Last reminder (local)',
+    )
+
+    @api.model
+    def _get_activities_to_remind_domain(self):
+        """Hook for extensions"""
+        return [('next_reminder', '<=', fields.Datetime.now())]
+
+    @api.model
+    def _get_activities_to_remind(self):
+        return self \
+            .search(self._get_activities_to_remind_domain()) \
+            .filtered(lambda activity: activity._should_be_reminded())
+
+    @api.model
+    def _process_reminders(self):
+        activities = self._get_activities_to_remind()
+        activities.action_remind()
+        return activities
+
+    @api.multi
+    def _should_be_reminded(self):
+        """Hook for extensions"""
+        self.ensure_one()
+        if self.next_reminder > fields.Datetime.now():
+            return False
+        tz = timezone(self.user_id.sudo().tz or 'UTC')
+        local_now = fields.Datetime.now().replace(tzinfo=UTC).astimezone(tz)
+        return self.date_deadline >= local_now.date()
+
+    @api.multi
+    @api.depends(
+        'user_id.tz',
+        'activity_type_id.reminders',
+        'date_deadline',
+        'last_reminder_local',
+    )
+    def _compute_next_reminder(self):
+        utc_now = fields.Datetime.now().replace(tzinfo=UTC)
+        for activity in self:
+            tz = timezone(activity.user_id.sudo().tz or 'UTC')
+            local_now = utc_now.astimezone(tz)
+            local_deadline = tz.localize(
+                datetime.combine(activity.date_deadline, time.min)
+            )
+            if local_deadline.date() < local_now.date():
+                activity.next_reminder = None
+                continue
+            reminders = activity.activity_type_id._get_reminder_offsets()
+            if not reminders:
+                activity.next_reminder = None
+                continue
+            reminders.sort(reverse=True)
+            last_reminder_local = tz.localize(
+                activity.last_reminder_local
+            ) if activity.last_reminder_local else None
+            for reminder in reminders:
+                next_reminder_local = local_deadline - relativedelta(
+                    days=reminder,
+                )
+                if not last_reminder_local \
+                        or next_reminder_local > last_reminder_local:
+                    break
+            if last_reminder_local \
+                    and next_reminder_local <= last_reminder_local:
+                activity.next_reminder = None
+                continue
+            activity.next_reminder = next_reminder_local \
+                .astimezone(UTC) \
+                .replace(tzinfo=None)
+
+    @api.multi
+    def action_notify(self):
+        super().action_notify()
+        utc_now = fields.Datetime.now().replace(tzinfo=UTC)
+        for activity in self.filtered(lambda x: not x.last_reminder_local):
```

Comment:
this way you loop twice. Check last_reminder_local in the loop and continue if not set

Reply:
Why in this case I loop twice?

Reply:
because filtered loops once then your loop does it again.
As a rule of thumb: never use filtered if you don't need the filtered recordset as-is and you don't loop on it right after.

Reply:
aah, @pedrobaeza usually said to move filtering to filtered if possible, if I got him correctly

Reply:
What I meant is to do `filtered` instead of list comprehensions or pure Python `filter`, but as @simahawk says, here you don't need to read whole recordset at the beginning if the loop is going to be interrupted and iterate twice. Any way, a performance note: if regular cases, whole recordset values are prefetched the same first time you accessed one field through dot notation, so that's not a game changer (or you explicitly put to not prefetch for improving that bits).

Reply:
so, how would it better in this case? move filtering inside loop?

Reply:
```py
        for activity in self:
            if activity.last_reminder_local:
                continue
```
as simple as this

Reply:
> Any way, a performance note: in regular cases, whole recordset values are prefetched the same first time you accessed one field through dot notation

If when `last_reminder_local` is first accessed, entire recordset of `activity` is fetched, and that's done for every item in self, what's the difference? O(n) here and O(n) there, same number of entries are accessed and thus prefetched.

Reply:
@alexey-pelykh The difference is that the filtering is done into the same loop as the rest of the logic...

`self.filtered(lambda x: not x.last_reminder_local)` -> 1 loop
` for activity in ` -> 1 loop
=> `for activity in self.filtered(lambda x: not x.last_reminder_local)`  -> 2 loops

With the @simahawk proposal you only have 1 loop: `for activity in self: `

Reply:
Done